### PR TITLE
Implement buffer block flattening

### PR DESCRIPTION
**Pull request by zeux**
Legacy GLSL targets do not support uniform buffers, and as such require
some sort of emulation. There are two alternatives - one is to represent
a uniform buffer as a uniform struct, and another one is to flatten it
into an array of primitive vector types (vec4).

Uniform struct have two disadvantages that make using them prohibitive
in some applications:

- The location assignment for struct members is arbitrary which means
the application has to set each struct member one by one
- Some Android drivers fail to link shader programs if both vertex and
fragment shader use the same uniform struct

Because of this, we need to support flattening uniform buffers into an
array. This is not just important for legacy GLSL but also is sometimes
useful for ESSL 3.0 where some Android drivers do not have stable UBO
support.

The way flattening works is the entire buffer is represented as a vec4
array; each access chain is rewritten into a combination of array
accesses, swizzles and data type constructors. Specifically:

- Extracting a vector or a scalar requires indexing into the array with
an optional swizzle, for example CB0[13].yz for reading vec2
- Extracting a matrix or a struct requires extracting each individual
vector or struct member and then combining them into the resulting
object
- Extracting arrays is not supported, mostly because the resulting
construct is very inefficient and ESSL 1.0 does not support array
constructors.

Additionally, while we try to constant-fold each individual indexing
operation, there are cases where we have to use dynamic index
computation (specifically for indexing arrays with non-constants); so
the general form of the primitive array extraction expression is:

    buffer[stride0*index0+...+strideN*indexN+offset]

Where stride/offset are integer literals and index represents variables.

### CODE REVIEW

**Comment by zeux** [comment]
This is a WIP change that addresses #79. I want to refactor & submit the part that moves type computation primitives from MSL to GLSL class separately, and the flattening code is missing column/row major matrix handling, base type verification and some other small details.

**Comment by HansKristian-ARM** [comment]
Looking good so far :) This is complex enough that we'll need some dedicated testing I think, so something like:
- In test_shader.py, check for .flatten. as part of the shader.
- Add a --flatten-buffer-blocks to main.cpp and use that in test_shader.py for shaders which contain .flatten.
- Add some dedicated test shaders for testing corner cases in legacy GLSL support.

I can do the plumbing here, but some test shaders with good coverage would be appreciated.

**Comment by zeux** [comment]
Yup, I'm planning to add tests - this is why I started looking at the test framework and setting up Travis build :)

**Comment by HansKristian-ARM** [comment]
Hi, what is the status of this PR?
I just added the default behavior of UBOs being emitted as uniform structs in https://github.com/KhronosGroup/SPIRV-Cross/pull/96 at least.

**Comment by zeux** [comment]
Didn't have a chance to work on this and then the holidays happened; I'll try to finish the remaining tasks soon. #96 looks like it would help this PR as well because of needs_transpose - which will make one the items I had yet to solve easier.

**Comment by HansKristian-ARM** [comment]
Cool, sounds good.

**Comment by zeux** [comment]
Looking at get_declared_type_size more closely, I now realize that there are two separate implementations of get_declared_struct_member_size - one in spirv_cross.cpp and one in spirv_msl.cpp that overrides the base impl. Looking at the code, it's very very close. I think the MSL version has a bug in get_declared_struct_member_size where it doesn't handle arrays at all.

I'm wondering why this is done this way. My inclination is to remove the virtual dispatch from get_declared_struct_member_size and just add get_declared_type_size to base Compiler class - that seems more consistent and eliminates a source of bugs. However, there aren't any tests for MSL so I can't really validate that this doesn't break anything, *and* there must have been a reason for this since AFAICT this virtual dispatch was introduced when MSL was introduced in the first place. Maybe @billhollings can shed some light on this/

**Comment by zeux** [comment]
Also looked a bit into row/column major handling. Doing it for matrices is pretty simple, unfortunately there are corner cases that are tricky - specifically, if you access a specific column (say, 1) in a matrix and the matrix is in a row-major format in a flattened buffer, what needs to happen is smth like:

    vec4(CB[0].y, CB[1].y, CB[2].y, CB[3].y)

or maybe (this doesn't work in GL2/GLES2):

    transpose(mat4(CB[0], CB[1], CB[2], CB[3]))[0]

Which isn't very easy in the current code structure - basically this has to be special cased (for access chains that go "through" a row major matrix but don't stop there). I'll see how much complexity this adds, there's always an option of detecting this and failing to compile the access chain for now.

**Comment by zeux** [comment]
I've added tests. I think this is ready modulo a) rebasing after #96 gets merged (which has multiple conflict points), b) figuring out what to do with MSL changes (this could be merged as is though), c) row-major support (currently flattening explicitly fails for row-major so it's a missing feature)

I'm currently considering doing b) & c) post-merge separately so removing the WIP tag.

**Comment by zeux** [comment]
When writing more tests, I've stumbled across one more issue with arrays. I believe now I understand the code structure a little more, but not the design decisions.

Let's say you have this uniform buffer:

    layout(std140) uniform UBO
    {
        vec4 Data[3][5];
    };

Here's how it's represented in SPIRV:

               6:             TypeFloat 32
               7:             TypeVector 6(float) 4
              13:             TypeInt 32 0
              14:     13(int) Constant 5
              15:             TypeArray 7(fvec4) 14
              16:     13(int) Constant 3
              17:             TypeArray 15 16
         18(UBO):             TypeStruct 17

Note that this has a simple recursive structure - UBO has one member which is an array of an array of a vector of floats. Every single item on this list is a separate id that can have metadata. Indeed, here's the relevant decorations from the shader:

                              Decorate 15 ArrayStride 16
                              Decorate 17 ArrayStride 80

So you can see that the array vec4[5] is annotated to have a stride of 16, and the array vec4[3][5] is annotated to have a stride of 80. It all makes sense.

Now unfortunately, SPIRV-Cross for some reason flattens out the array structure (I initially assumed SPIRV uses this type of specification, but it's not the case - it's just an artificial construct from SPIRV-Cross) with this code in Compiler::parse:

	case OpTypeArray:
	{
		uint32_t id = ops[0];

		auto &base = get<SPIRType>(ops[1]);
		auto &arraybase = set<SPIRType>(id);

		arraybase = base;

		auto *c = maybe_get<SPIRConstant>(ops[2]);
		bool literal = c && !c->specialization;

		arraybase.array_size_literal.push_back(literal);
		arraybase.array.push_back(literal ? c->scalar() : ops[2]);
		// Do NOT set arraybase.self!
		break;

As a result the struct member points (via an indirection) to vec4 directly, thereby resulting in the following structure:

        18: struct type, members: [17]
        17: vec4 type (id 7), array size: [3, 5]

Note that "15" is completely lost - it's now impossible to get the stride of the inner array. The function get_declared_type_size I copied from MSL source also doesn't work for much the same reasons. Technically maybe it shouldn't, since the type wasn't declared. Additionally, there's very important difference between "member_types" and "self" - specifically, if you get a type by id 17, self will be 7! Also the link between matrix->vector and vector->scalar types is destroyed as well, but at least it's not all flattened into "float"...

Now that I understand all this I can conclude that this structure is what led to most of the pain while implementing this feature. You can also see this structure poisoning the rest of the code - e.g. access_chain (and now flattened_access_chain) has this fun way to try to change the types while iterating to reconstruct the recursive structure that *was* there in the first place and got erased by the SPIRV parser. Similarly, in SPIRV-Cross there's a distinction between types that were declared and types that weren't, and querying information from struct members that doesn't work on non-struct-members - SPIRV does not have this but if you try to use the information from SPIRV (via type->self) you may find it was lost.

I've found a way to work around my issue which also means I can revert the changes I made to MSL, but overall all of this is far from ideal.

**Comment by HansKristian-ARM** [comment]
Ah, ok, I think I see what you mean. Yes, so the main reason for why the "self" and the type ID do not match, is that SPIR-V decorations and the type ID are kind of decoupled.

Only the base type (e.g. a plain OpTypeStruct) will have all the reflection information like offsets, names, etc, while the qualifiers like pointers/array did not add any significant information. To avoid a lot of special cases scattered around for qualified types vs plain types, I opted for this style from the beginning.

Up till now, only the outermost array stride was interesting, which could be fetched from the type ID itself. Unfortunately, all information about which type IDs compose the array type are lost, so array strides for partial arrays are not possible to get atm, I didn't consider that scenario.

I think it does make sense to add this information to SPIRType. Along with the "array" list, we should probably store the type IDs which ultimately compose the final array type so we can perform array stride reflection on them, that should help a bit when traversing the access chain.

I will look at your other PR first however.

**Comment by HansKristian-ARM** [comment]
Merged the conflicting PR, should I review this after you've resolved merge conflict?

File: spirv_glsl.cpp [hunk 1]
```diff
@@ -1037,6 +1037,17 @@ void CompilerGLSL::emit_buffer_block(const SPIRVariable &var)
...
 
+void CompilerGLSL::emit_flattened_buffer_block(const SPIRVariable &var)
+{
+	auto &type = get<SPIRType>(var.basetype);
+
+	// Block names should never alias.
+	auto buffer_name = to_name(type.self, false);
+	size_t buffer_size = (get_declared_struct_size(type) + 15) / 16;
+
+	statement("uniform vec4 ", buffer_name, "[", buffer_size, "];");
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: This doesn't handle flattened blocks of int/uint, but this can be done after the PR.

File: spirv_glsl.cpp [hunk 2]
```diff
@@ -3221,6 +3239,199 @@ string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32
...
+
+		return expr;
+	}
+	else
+	{
+		return access_chain(base, indices, count, false);
+	}
+}
+
+void CompilerGLSL::flattened_access_chain(std::string &expr, uint32_t base, const uint32_t *indices, uint32_t count, const SPIRType &target_type, uint32_t offset)
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: This could probably return a string instead to be more in line with the rest of the methods.

File: spirv_glsl.cpp [hunk 3]
```diff
@@ -3221,6 +3239,199 @@ string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32
...
+{
+	if (!target_type.array.empty())
+	{
+		SPIRV_CROSS_THROW("Access chains that result in an array can not be flattened");
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: I suppose that access chains of access chains in general won't easily work here, right?

File: spirv_glsl.cpp [hunk 4]
```diff
@@ -3221,6 +3239,199 @@ string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32
...
+
+void CompilerGLSL::flattened_access_chain_struct(std::string &expr, uint32_t base, const uint32_t *indices, uint32_t count, const SPIRType &target_type, uint32_t offset)
+{
+	expr += type_to_glsl(target_type);
+	expr += "(";
+
+	for (size_t i = 0; i < target_type.member_types.size(); ++i)
+	{
+		if (i != 0)
+			expr += ",";
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: ", ", should be a space after the comma.

File: spirv_glsl.cpp (hunk 1 above)
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: We could also check if the buffer block is read only, flattening SSBOs is just not going to happen :)

File: spirv_glsl.cpp [hunk 5]
```diff
@@ -3221,6 +3239,199 @@ string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32
...
+
+void CompilerGLSL::flattened_access_chain_matrix(std::string &expr, uint32_t base, const uint32_t *indices, uint32_t count, const SPIRType &target_type, uint32_t offset)
+{
+	expr += type_to_glsl(target_type);
+	expr += "(";
+
+	for (uint32_t i = 0; i < target_type.columns; ++i)
+	{
+		if (i != 0)
+			expr += ",";
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Space after comma.

File: spirv_glsl.cpp [hunk 6]
```diff
@@ -3221,6 +3239,199 @@ string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32
...
+
+		flattened_access_chain_vector_scalar(expr, base, indices, count, target_type, offset + i * 16);
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: There is a MatrixStride decoration for stride between columns. I think 16 would be correct here, but probably a good idea to not hardcode it.

File: spirv_glsl.cpp [hunk 7]
```diff
@@ -3221,6 +3239,199 @@ string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32
...
+
+	for (uint32_t i = 0; i < count; i++)
+	{
+		uint32_t index = indices[i];
+
+		// Arrays
+		if (!type->array.empty())
+		{
+			// We have to modify the type, so keep a local copy.
+			if (&temp != type)
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Ok, so instead of this hack, it would probably make sense to save the type of the "previous" array type in OpTypeArray to fix this hack.

```
type = get<SPIRType>(type->previous_array_type);
```
or something similar.

File: spirv_glsl.cpp [hunk 8]
```diff
@@ -3221,6 +3239,199 @@ string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32
...
+			temp.array.pop_back();
+
+			assert(type_size > 0);
+			assert(type_size % array_size == 0);
+
+			uint32_t array_stride = type_size / array_size;
+			assert(array_stride % 16 == 0);
+
+			expr += to_expression(index);
+			expr += "*";
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Spaces.

File: spirv_glsl.cpp [hunk 9]
```diff
@@ -3221,6 +3239,199 @@ string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32
...
+			expr += convert_to_string(array_stride / 16);
+			expr += "+";
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Spaces

File: spirv_glsl.cpp [hunk 10]
```diff
@@ -3221,6 +3239,199 @@ string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32
...
+		// Matrix -> Vector
+		else if (type->columns > 1)
+		{
+			if (ids[index].get_type() == TypeConstant)
+			{
+				auto &c = get<SPIRConstant>(index);
+				offset += c.scalar() * 16;
+			}
+			else
+				SPIRV_CROSS_THROW("Cannot flatten dynamic matrix indexing!");
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Isn't dynamically indexing here very similar to arrays?

File: spirv_glsl.cpp [hunk 11]
```diff
@@ -3221,6 +3239,199 @@ string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32
...
+		// Vector -> Scalar
+		else if (type->vecsize > 1)
+		{
+			if (ids[index].get_type() == TypeConstant)
+			{
+				auto &c = get<SPIRConstant>(index);
+				offset += c.scalar() * 4;
+			}
+			else
+				SPIRV_CROSS_THROW("Cannot flatten dynamic vector indexing!");
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Isn't dynamically indexing here very similar to arrays?

File: spirv_glsl.cpp [hunk 12]
```diff
@@ -3221,6 +3239,199 @@ string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32
...
+
+			// We have to modify the type, so keep a local copy.
+			if (&temp != type)
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Hm, seeing a pattern here, might make sense to just store the "previous" type. OpTypeVector also has a reference to the scalar type, OpTypeMatrix has a reference to the vector type, etc, etc ...

File: spirv_glsl.cpp [hunk 13]
```diff
@@ -5331,6 +5542,32 @@ void CompilerGLSL::require_extension(const string &ext)
...
+	auto &var = get<SPIRVariable>(id);
+	auto &type = get<SPIRType>(var.basetype);
+	auto name = to_name(type.self, false);
+	auto flags = meta.at(type.self).decoration.decoration_flags;
+
+	if (!type.array.empty())
+		SPIRV_CROSS_THROW(name + " is an array of UBOs.");
+	if (type.basetype != SPIRType::Struct)
+		SPIRV_CROSS_THROW(name + " is not a struct.");
+	if ((flags & (1ull << DecorationBlock)) == 0)
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Ok, seems like the UBO vs SSBO check is done here.

File: spirv_glsl.cpp [hunk 14]
```diff
@@ -5331,6 +5542,32 @@ void CompilerGLSL::require_extension(const string &ext)
...
+		SPIRV_CROSS_THROW(name + " is not a block.");
+	if (type.member_types.empty())
+		SPIRV_CROSS_THROW(name + " is an empty struct.");
+
+	uint64_t member_flags = 0;
+	for (uint32_t i = 0; i < type.member_types.size(); i++)
+		member_flags |= combined_decoration_for_member(type, i);
+
+	if (member_flags & (1ull << DecorationRowMajor))
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: We could probably support this if the access chain terminates at the matrix, the access chain expression can just get "need_transpose" in that case.

File: spirv_glsl.cpp (hunk 3 above)
Comment (zeux) [inline spirv_glsl.cpp]: The major limitation here actually is just that legacy GLSL (at least on GLES I think?) does not support array literals - the natural translation for something like this is `Light[4](Light(UBO[0].xyz, ...`. It's probably possible to work around this if there is a way to emit extra variables but TBH I haven't really seen a shader that copies an array out of the uniform buffers so the value is questionable...

File: spirv_glsl.cpp (hunk 12 above)
Comment (zeux) [inline spirv_glsl.cpp]: Yup - if this was the case it would be possible to basically reconstruct the original SPIRV hierarchy of types which would make this simpler because access chains traverse this hierarchy.

File: spirv_glsl.cpp (hunk 14 above)
Comment (zeux) [inline spirv_glsl.cpp]: Yeah - I'll need to do this because apparently glslang defaults to row major matrices in HLSL and flips the shader multiplication order, so this PR as it is only works on GLSL sources.

File: spirv_glsl.cpp (hunk 10 above)
Comment (zeux) [inline spirv_glsl.cpp]: Yeah, I guess this can work. Not sure how important this is - indexing matrices with non-literals may not be very common...

File: spirv_glsl.cpp (hunk 11 above)
Comment (zeux) [inline spirv_glsl.cpp]: Vector indexing is translated to swizzles so this has to be compile-time. It's theoretically possible to synthesize complex expressions such as `dot(UBO[4], lookUpTable[i])` with `vec4 lookUpTable[4]` but it seems like an overkill.

Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Dynamic vector lookup is possible in GLSL though, vec4 a; float b = a[index], but might complicate things a lot?

File: spirv_glsl.cpp (hunk 10 above)
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Not important enough to block a PR.

File: spirv_glsl.cpp (hunk 3 above)
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: iirc, glslang will actually copy out the array element by element anyways in this case as the storage types and "stack" types are different and aliasing is not allowed.

File: spirv_glsl.cpp (hunk 6 above)
Comment (zeux) [inline spirv_glsl.cpp]: 16 is basically the only valid value for std140 (GL spec 7.6.2.2 Standard Uniform Block Layout), and the code assumes this in some other places, specifically for array elements. The problem with handling other strides (say, for std430) is that it can result in vectors being "split" across individual elements, which is not supported by the code right now - so it's not enough to just use another value here.

File: spirv_glsl.cpp (hunk 11 above)
Comment (zeux) [inline spirv_glsl.cpp]: Looks like ESSL 1.0 is undecided:

> Should dynamic indexing of vectors and matrices be removed from the specification?
> RESOLUTION: Keep in main specification. Support is not mandated.

I'm, uh, not sure what this means exactly :) "use at your own peril" maybe.

Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Uh ... ye, nvm then. :p

File: spirv_glsl.cpp (hunk 2 above)
Comment (zeux) [inline spirv_glsl.cpp]: I was mostly wary of excessive string copies, and I'd need to return smth like a std::pair from flattened_access_chain_offset. I can do that though if this is more consistent with the rest of the code.

**Comment by zeux** [comment]
Updated with whitespace fixes, code cleanup and row-major support for access chains that end in a matrix. Extra extensions after this PR could include:

- Support for int buffer flattening
- Support for dynamic matrix indexing
- Support for row major matrix indexing (that is, access chains with row-major matrices in the middle)

From my perspective only the last item is interesting (I don't have use cases for the first two); I mocked up an implementation for and it's pretty simple but it can probably be in a separate PR. Please let me know if any other revisions are necessary!

File: spirv_glsl.cpp-f102ece2 [hunk 15]
```diff
@@ -0,0 +1,6394 @@
+/*
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp-f102ece2]: Commited a temp file?

File: spirv_glsl.cpp [hunk 16]
```diff
@@ -3252,6 +3290,208 @@ string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32
...
+	expr += result.first; // this is a series of N1*k1+N2*k2+... that is either empty or ends with a +
+	expr += convert_to_string(index / 4);
+	expr += "]";
+
+	expr += vector_swizzle(target_type.vecsize, index % 4);
+
+	return expr;
+}
+
+std::pair<std::string, uint32_t> CompilerGLSL::flattened_access_chain_offset(uint32_t base, const uint32_t *indices, uint32_t count, uint32_t offset, bool *need_transpose)
```
Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Should have clarified, for these functions which returned offset, the expr as a param was fine, it was the access chain ones which returned void which could return string.

File: spirv_glsl.cpp-f102ece2 (hunk 15 above)
Comment (zeux) [inline spirv_glsl.cpp-f102ece2]: Oops. Too much rebasing. Removed and updated.

File: spirv_glsl.cpp (hunk 16 above)
Comment (zeux) [inline spirv_glsl.cpp]: It's actually a bit nicer that way since I can easily reuse flattened_access_chain_offset in an arbitrary location at the call site and then use the resulting expression to make the final one later.

Comment (HansKristian-ARM) [inline spirv_glsl.cpp]: Ah, ok, let's keep it then.

**Comment by HansKristian-ARM** [comment]
Thanks for this PR.
